pkg/scheduler/queue/query_component_capacity.go

```diff
@@ -0,0 +1,179 @@
+
+const MinReservedQueryComponentCapacity = 0.1
+const MaxReservedQueryComponentCapacity = 0.5
+
+func NewQueryComponentCapacity(
+	targetReservedCapacity float64,
+	querierInflightRequests *prometheus.GaugeVec,
+) (*QueryComponentCapacity, error) {
+
+	if targetReservedCapacity < MinReservedQueryComponentCapacity || targetReservedCapacity >= MaxReservedQueryComponentCapacity {
+		return nil, errors.New("invalid targetReservedCapacity")
+	}
+
+	return &QueryComponentCapacity{
+		targetReservedCapacity: targetReservedCapacity,
+
+		ingesterInflightRequests:     atomic.NewInt64(0),
+		storeGatewayInflightRequests: atomic.NewInt64(0),
+		querierInflightRequestsTotal: atomic.NewInt64(0),
+
+		querierInflightRequestsGauge: querierInflightRequests,
+	}, nil
+}
+
+// ExceedsCapacityForComponentName checks whether a query component has exceeded the capacity utilization threshold.
+// This enables the dequeuing algorithm to skip requests for a query component experiencing heavy load,
+// reserving querier-worker connection capacity to continue servicing requests for the other component.
+// If there are no requests in queue for the other, less-utilized component, the dequeuing algorithm may choose
+// to continue servicing requests for the component which has exceeded the reserved capacity threshold.
+//
+// Capacity utilization for a QueryComponent is defined by the portion of the querier-worker connections
+// which are currently in flight processing a query which requires that QueryComponent.
+//
+// The component name can indicate the usage of one or both of ingesters and store-gateways,
+// and will be indicated as exceeding the threshold if either ingesters or store-gateways
+// are currently in excess of the reserved capacity.
+//
+// The threshold for a QueryComponent's utilization of querier-worker connections
+// can only be exceeded by one QueryComponent at a time as long as targetReservedCapacity is < 0.5.
+// Therefore, one of the components will always be given the OK to dequeue queries for.
+//
+// Capacity reservation only occurs when the queue backlogged, where backlogged is defined as
+// (length of the query queue) >= (number of querier-worker connections waiting for a query).
+//
+// A QueryComponent's utilization is allowed to exceed the reserved capacity when the queue is not backlogged.
+// If an influx of queries then creates a backlog, this method will indicate to skip queries for the component.
+// As the inflight queries complete or fail, the component's utilization will naturally decrease.
+// This method will continue to indicate to skip queries for the component until it is back under the threshold.
+func (qcl *QueryComponentCapacity) ExceedsCapacityForComponentName(
+	name string, connectedWorkers *atomic.Int64, queueLen, waitingWorkers int,
+) (bool, QueryComponent) {
+	if waitingWorkers > queueLen {
+		// excess querier-worker capacity; no need to reserve any for now
+		return false, ""
```

Comment:
Do we need both the boolean and string return value here? Seems like we just need to know whether we should or shouldn't avoid this query request.

Reply:
it's useful for tests and could be included in the logging if we wanted.